File: trunk/mobileCloud/android/2_0/native-framework/src/main/java/org/openmobster/core/mobileCloud/android_native/framework/ExecuteOnEDT.java
/**
 * Copyright (c) {2003,2010} {openmobster@gmail.com} {individual contributors as indicated by the @authors tag}.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 */
package org.openmobster.core.mobileCloud.android_native.framework;

import org.openmobster.core.mobileCloud.android.errors.ErrorHandler;
import org.openmobster.core.mobileCloud.android.errors.SystemException;
import org.openmobster.core.mobileCloud.api.ui.framework.Services;
import org.openmobster.core.mobileCloud.api.ui.framework.command.AppException;
import org.openmobster.core.mobileCloud.api.ui.framework.command.Command;
import org.openmobster.core.mobileCloud.api.ui.framework.command.CommandContext;
import org.openmobster.core.mobileCloud.api.ui.framework.command.CommandService;
import org.openmobster.core.mobileCloud.api.ui.framework.command.LocalCommand;
import org.openmobster.core.mobileCloud.api.ui.framework.command.RemoteCommand;

/**
 * @author openmobster@gmail.com
 *
 */
final class ExecuteOnEDT implements Runnable
{
	private CommandContext commandContext;
	
	ExecuteOnEDT(CommandContext commandContext)
	{
		this.commandContext = commandContext;
	}
	
	public void run()
	{
		try
		{
			CommandService service = Services.getInstance().getCommandService();			
			Command command = service.find(commandContext.getTarget());
			
			if(command instanceof LocalCommand)
			{
				//View Before
				command.doViewBefore(commandContext);				
				
				//Perform action
				try
				{
					command.doAction(commandContext);
				}
				catch(AppException ape)
				{
					service.reportAppException(commandContext, ape);
					command.doViewError(commandContext);
					return;
				}
												
				//View After
				command.doViewAfter(commandContext);
			}
			else if(command instanceof RemoteCommand)
			{
				//View Before
				command.doViewBefore(commandContext);
				
				//Perform this in async mode
				new LongAsyncTask().execute(commandContext);
			}
		}
		catch(Exception e)
		{
			//report to ErrorHandling system
			ErrorHandler.getInstance().handle(new SystemException(this.getClass().getName(), "execute", new Object[]{
				"Message:"+e.getMessage(),
				"Exception:"+e.toString(),
				"Target Command:"+commandContext.getTarget()
			}));
			ShowError.showGenericError(commandContext);
		}
	}
}

File: iphone/mobilecloudlib/ExecuteOnEDT.m
/**
 * Copyright (c) {2003,2010} {openmobster@gmail.com} {individual contributors as indicated by the @authors tag}.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 */

#import "ExecuteOnEDT.h"


/**
 * 
 * @author openmobster@gmail.com
 */
@implementation ExecuteOnEDT

@synthesize commandContext;

-(id)initWithContext:(CommandContext *)context
{
	if([super init] == self)
	{
		queue = [[NSOperationQueue alloc] init];
		self.commandContext = context;
	}
	return self;
}

-(void)dealloc
{
	[queue release];
	[commandContext release];
	if(activityIndicatorView != nil)
	{
		[activityIndicatorView release];
	}
	[super dealloc];
}

+(id)withInit:(CommandContext *)commandContext
{
	ExecuteOnEDT *instance = [[ExecuteOnEDT alloc] initWithContext:commandContext];
	instance = [instance autorelease];
	
	return instance;
}

-(void)run
{
	@try 
	{
		[self startBusyCommand];
	}
	@catch (NSException * e) 
	{
		//FIXME: If this happens show a Global Error Message
	}
}

-(void)startBusyCommand
{	
	//Start the Activity Indicators
	activityIndicatorView = [[ActivityIndicatorView alloc] initWithNibName:@"ActivityIndicatorView" bundle:nil];
	UIViewController *caller = commandContext.caller;
	
	//Disable the screen from interactions
	[caller.view setUserInteractionEnabled:NO];
	if(caller.navigationController != nil)
	{
		[caller.navigationController.view setUserInteractionEnabled:NO];
	}
	if(caller.parentViewController != nil)
	{
		[caller.parentViewController.view setUserInteractionEnabled:NO];
	}
	if(caller.tabBarController != nil)
	{
		[caller.tabBarController.view setUserInteractionEnabled:NO];
	}
	
	[caller.view addSubview:activityIndicatorView.view];
	[caller.view bringSubviewToFront:activityIndicatorView.view];
	
	NSInvocationOperation *operation = [[NSInvocationOperation alloc] initWithTarget:self selector:@selector(invokeBusyAction) object:nil];	
	operation = [operation autorelease];
	
	//Add the operation to the queue
	[queue addOperation:operation];
}

-(void)invokeBusyAction
{
	@try
	{
		id<Command> target = (id<Command>)[commandContext getTarget];
		[target doAction:commandContext];
	}
	@catch(SystemException *syse)
	{
		[commandContext setError:@"500" :[syse getMessage]];
	}
	@catch (NSException *e) 
	{
		NSString *message = @"No Message Found";
		if([e reason] != nil)
		{
			message = [e reason];
		}
		else if([e name] != nil)
		{
			message = [e name];
		}
		[commandContext setError:@"500" :message];	
	}
	[self performSelectorOnMainThread:@selector(endBusyCommand) withObject:nil waitUntilDone:NO];
}

-(void)endBusyCommand
{
	@try 
	{
		//Stop the Activity Indicator
		[activityIndicatorView.view removeFromSuperview];
		UIViewController *caller = commandContext.caller;
	
		if(caller.navigationController != nil)
		{
			[caller.navigationController.view setUserInteractionEnabled:YES];
		}
		if(caller.parentViewController != nil)
		{
			[caller.parentViewController.view setUserInteractionEnabled:YES];
		}
		if(caller.tabBarController != nil)
		{
			[caller.tabBarController.view setUserInteractionEnabled:YES];
		}
		[caller.view setUserInteractionEnabled:YES];
	
		if([caller conformsToProtocol:@protocol(UICommandDelegate)])
		{
			UIViewController<UICommandDelegate> *local = (UIViewController<UICommandDelegate> *)caller;
			
			if(![commandContext hasErrors])
			{
				[local doViewAfter:commandContext];
			}
			else 
			{
				[local doViewError:commandContext];
			}
		}
	}
	@catch (NSException * e) 
	{
		//FIXME: If this happens show a Global Error Message
	}
}
@end
